### PR TITLE
Fix nuts step size adaptation

### LINKED ISSUE #1697: Step size adaptation in NUTS
The algorithm implemented for step size adaptation in pymc3 differs from the algorithm described in the paper about nuts and from the stan implementation. The algorithm in the paper (algorithm 6) adaptes epsilon very quickly for the first 1000 iterations, while keeping some average epsilon around. After these 1000 iterations it stops the adaptation and only uses the average epsilon. pymc3 sticks to the first epsilon all the time, and also uses a slightly different way to update it (`log(e)' = mu - m ** k / gamma H_bar` instead of `log(e) = mu - sqrt(m) / gamma H_bar`). Is this intentional? The docstring doesn't say anything about this.

### PR DESCRIPTION
**Pull request by aseyboldt**
This fixes #1697 
At the moment this is a bit difficult to test, as nuts doesn't expose its state in any way. I can add a few more detailed tests when #1687 is finished.
There should probably also be some updates to the documentation, as this basically introduces a new parameter for nuts. The method of calling nuts that is used in the doc doesn't use the new step size at all. I think this:
```python
with model:
    trace = pymc3.sample(2000)
```
should probably be replaced by this:
```python
with model:
    trace = pymc3.sample(2000, tune=1000)
    trace = trace[1000:]
```

### CODE REVIEW

**Comment by fonnesbeck** [comment]
That's been suggested before, and the preference (my preference, at least) was to take care of burn-in during post-processing. I would be on board with this if this was automatic--that is, rather than specify a guess at the tuning interval, do not start recording trace values until NUTS is finished tuning. This would require some logic to determine that the step size is tuned. You could take it a step further and use Rhat with multiple chains to not begin recording the sample until it is less than 1.1 for all variables.

**Comment by jsalvatier** [comment]
@fonnesbeck the change here is not the `tune` parameter, but to change the step size adaptation from a Robins and Monroe scheme to a 'primal-dual' scheme. Like it was supposed to be all along. See the nuts [paper](http://www.stat.columbia.edu/~gelman/research/published/nuts.pdf) section 3.2.1.

**Comment by twiecki** [comment]
@jsalvatier did you review the changes / math here?

**Comment by twiecki** [comment]
```
======================================================================
FAIL: test_posterior_estimate (pymc3.tests.test_step.TestSampleEstimates)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_step.py", line 281, in test_posterior_estimate
    np.testing.assert_array_less(0.05, p_alpha, verbose=True)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 990, in assert_array_less
    header='Arrays are not less-ordered')
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 739, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not less-ordered
(mismatch 100.0%)
 x: array(0.05)
 y: array(0.04227803959562242)
-------------------- >> begin captured logging << --------------------
pymc3: DEBUG: Applied log-transform to sigma and added transformed sigma_log_ to model.
--------------------- >> end captured logging << ---------------------
======================================================================
FAIL: pymc3.tests.test_step.TestStepMethods.test_sample_exact(<class 'pymc3.step_methods.hmc.nuts.NUTS'>,)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_step.py", line 136, in check_trace
    assert_array_almost_equal(trace.get_values('x'), self.master_samples[step_method])
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 918, in assert_array_almost_equal
    precision=decimal)
  File "/home/travis/miniconda3/envs/testenv/lib/python3.6/site-packages/numpy/testing/utils.py", line 739, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 6 decimals
(mismatch 98.0%)
 x: array([ 1.118324,  1.118324,  0.428752, -1.063373, -1.175014, -0.277752,
       -0.277752,  0.003286,  0.23796 ,  1.655296,  1.790141, -0.623905,
        0.60124 ,  0.60124 ,  0.684407,  0.704353,  0.602374,  0.037173,...
 y: array([ 1.118324,  1.118324,  0.629616, -1.887259, -0.280856, -0.280856,
        0.512465,  0.512465,  0.515242, -1.074791, -1.429564, -1.429564,
        0.236747, -1.047215, -0.971671, -0.971671, -0.659039,  1.392075,...
```

@aseyboldt Is the implementation supposed to change behavior if tuning=False?

**Comment by fonnesbeck** [comment]
@jsalvatier yes, I was referring to the suggestion to add a tune interval argument explicitly above

**Comment by aseyboldt** [comment]
@twiecki That's actually nice to see :-). I just ran the tests with `nosetests . ` and got no failure, which took me a bit by surprise. The epsilon update changed from using `m ** k` to `sqrt(m)` (k had a default value of 0.75), even while tune = False. k is now only used in the update of epsilon_bar. I'll fix the test failures tomorrow. Although the difference between x and y looks like it might be a bit large...

@fonnesbeck Sorry, the description in this pull request was somewhat ambiguous. NUTS, as described in the paper reference by jalvatier above, needs an additional parameter: the number of steps it should use to find a good step size. Apparently this was left out by mistake. This pull request passes that parameter to nuts using the tune parameter of the sampler, so after `tune` many steps, *the behaviour of NUTS changes*. The comment about changing the docs was not about how to deal with the usual burn in, but about telling users, that the interface for using NUTS changed.

Also, how should NUTS behave if no `tune` parameter is set? With this pull request that would be accepted without complaint, but just never switch to the final step size.

File: pymc3/step_methods/hmc/nuts.py [hunk 1]
```diff
@@ -50,13 +53,22 @@ def __init__(self, vars=None, Emax=1000, target_accept=0.8,
         self.t0 = t0
 
         self.h_bar = 0
-        self.u = np.log(self.step_size * 10)
+        self.mu = np.log(self.step_size * 10)
+        self.lstep_size = np.log(self.step_size)
```
Comment (ColCarroll) [inline pymc3/step_methods/hmc/nuts.py]: nit: can this be `self.log_step_size` and `self.log_step_size_bar`?

**Review by ColCarroll — COMMENTED** [review]
Thanks a lot for this -- especially some of the additional exposition in the doc! 

This matches the API for other sampling methods w.r.t. the `tune` parameter (i.e., user is responsible for discarding unwanted samples), so I'm not sure much more work is needed, other than updating the `exact_sample` test, and perhaps some extra hard looks at this not introducing unexpected problems!

File: pymc3/step_methods/hmc/nuts.py (hunk 1 above)
Comment (aseyboldt) [reply pymc3/step_methods/hmc/nuts.py]: done

**Pull request merged**